koku/masu/util/common.py

```diff
@@ -325,52 +326,74 @@ def batch(iterable, start=0, stop=None, _slice=1):
+    LOG.info(log_json(msg="checking tags should be enabled", context=ctx))
+    if not enabled_tags:
+        LOG.info(log_json(msg="skipping tag enablement no tags found", context=ctx))
         return
 
-    LOG.info(log_json(msg="creating enabled key records", context=ctx))
-    changed = False
-
     with schema_context(schema):
-        new_keys = list(set(enabled_keys) - {k for k in enabled_keys_model.objects.values_list("key", flat=True)})
-        if new_keys:
-            changed = True
-            # Processing in batches for increased efficiency
-            for batch_num, new_batch in enumerate(batch(new_keys, _slice=500)):
-                batch_size = len(new_batch)
-                LOG.info(
-                    log_json(msg="create batch", batch_number=(batch_num + 1), batch_size=batch_size, context=ctx)
-                )
-                for ix in range(batch_size):
-                    new_batch[ix] = enabled_keys_model(key=new_batch[ix])
-                enabled_keys_model.objects.bulk_create(new_batch, ignore_conflicts=True)
-    if not changed:
-        LOG.info(log_json(msg="no enabled keys added", context=ctx))
-
-    return changed
-
+        new_tags = list(
+            set(enabled_tags)
+            - {k for k in EnabledTagKeys.objects.filter(provider_type=provider_type).values_list("key", flat=True)}
+        )
+        if not new_tags:
+            LOG.info(log_json(msg="no new tags founds skipping tag enablement", context=ctx))
+            return
+
+        for batch_num, new_batch in enumerate(batch(new_tags, _slice=500)):
+            batch_size = len(new_batch)
+            if enabled_value:
+                ctx["enabled_tag_limit"] = Config.ENABLED_TAG_LIMIT
+                if enabled_tag_count == 0:
+                    LOG.debug("querying for enabled tag count.")
+                    enabled_tag_count = EnabledTagKeys.objects.filter(enabled=True).count()
+                if Config.ENABLED_TAG_LIMIT <= 0:
+                    # If the tag limit is zero or less, the limit is disabled.
+                    LOG.info(log_json(msg="tag limit is disabled", context=ctx))
+                elif enabled_tag_count < Config.ENABLED_TAG_LIMIT:
+                    LOG.info(log_json(msg="tag limit is enabled", context=ctx))
+                else:
+                    LOG.info(log_json(msg="tag limit has been reached"), content=ctx)
+                    enabled_value = False
+            ctx["tags_enabled"] = enabled_value
+            batch_size = len(new_batch)
+            LOG.info(
+                log_json(msg="create tag batch", batch_number=(batch_num + 1), batch_size=batch_size, context=ctx)
+            )
+            for ix in range(batch_size):
+                new_batch[ix] = EnabledTagKeys(key=new_batch[ix], provider_type=provider_type, enabled=enabled_value)
+            enabled_tag_count += batch_size
+            EnabledTagKeys.objects.bulk_create(new_batch, ignore_conflicts=True)
```

Comment:
This is modifying the list while iterating over it. That works fine here because the size of the list isn't changing, but it would be safer to build a different object to pass to `bulk_create()`.
```suggestion
            new_records = (EnabledTagKeys(key=key, provider_type=provider_type, enabled=enabled_value) for key in new_batch)
            enabled_tag_count += batch_size
            EnabledTagKeys.objects.bulk_create(new_records, ignore_conflicts=True)
```